server-common/src/main/java/org/apache/kafka/server/immutable/ImmutableMap.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server.immutable;
+
+import java.util.Map;
+
+/**
+ * A persistent Hash-based Map wrapper.
+ * java.util.Map methods that mutate in-place will throw UnsupportedOperationException
+ *
+ * @param <K> the key type
+ * @param <V> the value type
+ */
+public interface ImmutableMap<K, V> extends Map<K, V> {
```

Comment:
Did we consider adding the methods to create collections here as static methods?

Reply:
If we add static methods here then those methods would have to hard-code which underlying library to use.  Right now we just have PCollections because it meets a mandatory requirement for the ACL lock removal work, but it is possible we might find either a need or a significant performance advantage to using something else for some use case.  Any static methods here might have to have the underlying library in the name.  For example, a method name here might be `emptyMapPCollections()` (or `pCollectionsEmptyMap()`.  I assumed it was easy for clients to just explicitly invoke `PCOLLECTIONS_FACTORY.emptySet()` (for example) -- or, more likely, they would have access to the factory already (like `TopicsImage` does as it declares `private static final ImmutableMapSetFactory FACTORY = ImmutableMapSetFactory.PCOLLECTIONS_FACTORY;`) and could invoke `FACTORY.emptySet()`.

Reply:
The whole point is abstracting things from users. I don't think they should be making the call and I don't think we should be including the name of the library. If we need to expose a collection from a different library because it has specific properties (say performance), then we'd use a name that makes _that_ clear.

Reply:
I don't feel strongly about it, but I agree that from a practical point of view it might be easier to have a factory method. Then you could just replace that factory method implementation with the Vavr implementation (or whatever) and rerun the JMH benchmarks without changing anything else.

Reply:
I updated it.  I removed the factory in favor of static methods on the interface.

Reply:
I think this looks pretty clean.